### ISSUE TITLE
Way to explicitly set observed on computed

### ISSUE THREAD

**Issue by scottmas** [mobxjs/mobx#1330]
I've looked and looked and cannot find any way that it is possible to explicitly set which observables should be tracked in a computed invocation. The best I've come up with is:
```javascript
@computed get prop(){
   console.log(this.someObservable) // Explicitly trigger the observables you want to be reacted to
   untracked(() => {
      //Do your stuff that potentially touches lots of observables. 
  })
}
```

But I don't love this approach since it feels a bit dirty. Is there no way to explicitly specify which observables I want a computed getter to react to?

**Comment by mweststrate** [comment]
@scottmas MobX is explicitly designed to _not_ do that :). Could you tell _why_ you would not want certain observables to be tracked?

**Comment by urugator** [comment]
The Mobx reasoning is that it has to subscribe for anything which may affect the result of computation. Anything else would cause staleness or oversubscription.
So you're either using MobX wrong way or it's some kind of workaround in which case it should look dirty ...

**Comment by scottmas** [comment]
Take the following code: 

```javascript
class Game{
  @observable gameEvents = [];
  @observable gameClock = 0; // Update this every 1000 ms
  @computed get activePlayers(){
       const activePlayers = [];
    
       this.gameEvents.forEach((event) => {
           // TODO: compute active players from game events
           activePlayers.push({playingTime: 2340}, new Player(this))
           // Oh no! activePlayers now listens for changes to the game clock
       })   
      
       return activePlayers;
   }
}

class Player{
   @computed get playerClock(){
      //Some computation involving the game clock and player clock
      return this.game.gameClock + this.playingTime 
   }
   constructor(playerObj, game){
       this.playingTime = playerObj.playingTime
       this.game = game;
    }
}

@observer
class SomeComponent extends React.Component{
  render(){
     // Oh no! This causes a recomputation of Game.activePlayers on each tick of the game clock.
     return (
        <div>{this.props.player.playingTime}</div>
     )
   }
}
```

The central problem is that I don't want to trigger a new computation of `activePlayers` on each tick of the game clock. But I do want to update components that are subscribed to an observable on `Player`. I know ahead of time that the only way `activePlayers` will ever change is if there is a new entry to `gameEvents`. I can accomplish my desired behavior with `untracked` like I said, but it feels messy. I know ahead of time what can cause a re-computation of `activePlayers`, and it's not the game clock, although `gameClock` needs to be an observable. 

Really, what I think would be nice is something conceptually similar to `componentShouldUpdate`, but for computed properties.

**Comment by urugator** [comment]
> // Oh no! activePlayers now listens for changes to the game clock

It doesn't, at least in this example, because `this.gameClock` isn't used anywhere during the computation...

**Comment by scottmas** [comment]
Sorry, yes you are right. Updated the example to use the `playerClock`, causing the undesired behavior.

**Comment by scottmas** [comment]
Obviously it's **possible** to do this with untracked. But I worry that future devs (myself included) will forget about the recomputation side effects and accidentally cause the exact situation I'm trying to avoid. It's like `shouldComponentUpdate` - yes it's a sledgehammer and you can kick yourself in the foot, but sometimes its the best tool for the job.

**Comment by scottmas** [comment]
A `computed.on` decorator is basically what I think I would like.

```javascript
@observable gameEvents = [];
@computed.on(this.gameEvents) get activePlayers(){
  //Do stuff
}

**Comment by urugator** [comment]
> doSomething(newPlayer.playerClock);

Please could you be more specific? The chances are that you should design things differently, but without knowing the intentions, it's hard to help.
`doSomething(newPlayer.playerClock);` would suggest that there are some side effects in computed which is forbidden.
I am not the one to decide, but `computed.on` or anything similar is unlikely to happen.

**Comment by scottmas** [comment]
In this particular case, it's actually not so much whether or not I can do it in a different way that doesn't trigger the listener, but rather that I want to avoid unwanted situations. Although "side effects in computed values" are forbidden, as you say, there's nothing that stops such observables from being accidentally observed if they are referenced in the getter, which is something I very much want to avoid on this expensive computation. 

Like I said, in my particular use case I know with 100% certainty what the getter should react to, and I'm more worried about accidentally triggering a bajillion recomputations than I am about accdientally forgetting to update the explicit lists of observables when a new requirements comes in. 

Wouldn't you say this use case is conceptually very similar to `shouldComponentUpdate` for React components? Is there anything so bad about having an escape hatch for when you want to optimize performance and total computations?

**Comment by urugator** [comment]
> side effects

By "side effects" I mean state modifications, it's probably not the case, the point is that it's impossible to tell from your example...

> Wouldn't you say this use case is conceptually very similar to shouldComponentUpdate for React components?

Actually the `equals` options is conceptually closer to `shouldComponentUpdate` inside MobX ecosystem.
In react the state is propagated from top to bottom (from parent to children)
In computed the state is propagated from bottom to top (from atom to computation to another computation to reaction...).
In react the parent component decides whether the children should be notified about state change by sending new props.
In computed the dependencies decides whether the dependant (computed) should be notified about state change by sending their new value.
In react the notified component uses `shoulComponentUpdate` to decide whether the state actually changed.
In mobx the notified computation uses `equals` to decide whether the state actually changed.

In other words `shouldComponentUpdate` doesn't decide whether the component should be notified about the possible change, but whether it should trigger the reaction (render).
If you would want to mimic this in Mobx, you would have to provide a function, which filters the computed's dependencies, something like `computed.shouldUpdate(dep => dep === gameClock ? false : true)`.
However this is obviously impossible, because dependencies are not known until the computation is executed.

What you're asking for is explicit subscription mechanism. There is nothing like that in React ... children component can't subscribe for specific parent state changes.

**Comment by scottmas** [comment]
Hmm, so is there a way to do what I want to do with the `equals` option?

**Comment by spion** [comment]
Basically the question is, do you need `doSomething(newPlayer.playerClock)` to compute the active players?

* If you don't, why are you calling it in the computed?
* If you do, does the computed not potentially change when the clock changes?

**Comment by scottmas** [comment]
You're right, this is definitely possible. And properly done, I think you're right that any of the referenced observables will be required in computing the new observable. 

I think it's more of a convenience thing. For instance, if I'm debugging and log a player, that will begin tracking of the unwanted observable. And there could be edge cases where the dev isn't thinking and references a computed variable that SHOULD not be referencing the unwanted observable. For instance, imagine a kitchen sink observable that among other things references the unwanted observable. Triggering the unwanted observable is purely accidental, but he doesn't notice he's done it because things continue working as expected. The only downside is that recomputations are now being triggered way too often, causing a slight performance degradation. 

I mean, this is definitely possible right now just using untracked, so maybe I just keep using that:
```
@computed get activePlayers(){
  this.gameEvents; //No-op to declare dependencies
  return untracked(() => {
     //Do stuff and return activePlayers
  })
}
```

**Comment by spion** [comment]
I would look into splitting your computed so that expensive sub-computations are cached separately. Using untracked will probably eventually lead to stale values.

**Comment by mweststrate** [comment]
Closing question as the issue has become inactive. Feel free to reopen if more follow up is needed

**Comment by iMoses** [comment]
I have another example in which I'm struggling to find a solution.

Consider this code:
```
@observable entities = new Map;
@observable dirtyList = [];  // contains a list of entities keys which are dirty

@computed get dirtyEntities() {
  return this.dirtyList.map(key => this.entities.get(key));
}

changeSomething(data) {
  transaction(() => {
    this.entities.merge(data);
    if (somethingIsNowDirty) {
      // make sure changes to dirtyList are always in the same transaction as entities
    }
  });
}
```

The entities are observable by themselves, so there's really no need to update `dirtyEntities` every time a change is made inside `entities`, but only once changes were made to `dirtyList`.

Any suggestions on how to handle such a use case?

**Comment by FredyC** [comment]
Perhaps you care to explain why you need `dirtyEntities` updated within a transaction? Sounds like another bad way of using MobX. You should setup a reaction instead if you need to do something based on observable change.

**Comment by iMoses** [comment]
> Perhaps you care to explain why you need `dirtyEntities` updated within a transaction? Sounds like another bad way of using MobX. You should setup a reaction instead if you need to do something based on observable change.

I don't understand your question, I don't need `dirtyEntities` updated within a transaction. 
The transaction is to ensure that changes to `entities` and `dirtyList` are made together, to prevent an event in which `dirtyEntities` is triggered before both values were updated.

**Comment by urugator** [comment]
@iMoses 

> update dirtyEntities every time a change is made inside entities

It won't update every time, but only if the accessed key is modified or deleted.

You can disable subscription for `this.entities` like this (use at own risk): 
```javascript
@computed get dirtyEntities() {
  return this.dirtyList.map(key => untracked(() => this.entities.get(key)));
}
```

**Comment by iMoses** [comment]
@urugator Thank you! That's very good to know..

What are the risks I should take into consideration? 
Anything other than `entities` not being tracked?

**Comment by urugator** [comment]
> Anything other than entities not being tracked?

Well no, but the whole point of this lib is automatic subscription management.
Once you start using `untracked` or similar, you get really convoluted manual subscription management.

**Comment by iMoses** [comment]
I simply see it as providing me with the tools to fine tune my
observations. I'm still getting an observable computed getter from the
library and that's fantastic IMO.
